### PR TITLE
Drop `Optional` and `OptionalViewState` classes

### PR DESCRIPTION
**Pull request by wzieba**
### Context
`Dependabot` created a PR with a bump to Kotlin `1.4.32` #3751  and it fails during the build. The reason is a bug in `kapt` that messes something with `@Parcelize` code generation. In our case, this scenario only appears for `OptionalViewState` class. More details here: https://youtrack.jetbrains.com/issue/KT-42652

Discussion: p1616589646117900-slack-CGPNUU63E

#### TC1 - variation sale end date
1. Ensure there's a product with variation
2. Go to variation details
2. Go to `Price` 
3. Schedule sale
4. Go back - ensure the `Sale dates` in `Price` area are setup correctly
5. Go to `Price`
6. Remove only `to` date from sale schedule
7. Go back - ensure that there's no sale end date in `Price` area

#### TC2 - variation image upload
1. Go to images section of the variation
2. Smoke test here: update photo, remove it, re-add. Make sure that there's an upload progress bar and that pictures are uploaded.

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailViewModel.kt [hunk 1]
```diff
@@ -711,7 +710,6 @@ class ProductDetailViewModel @AssistedInject constructor(
         attributes: List<ProductAttribute>? = null
     ) {
         viewState.productDraft?.let { product ->
-            val currentProduct = product.copy()
```
Comment (wzieba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailViewModel.kt:714]: `currentProduct` has the same value as `product` (even in `Product#copy` as its before evaluation of `copy` method)

File: WooCommerce/src/test/kotlin/com/woocommerce/android/ui/products/variations/VariationDetailViewModelTest.kt [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+import com.woocommerce.android.viewmodel.SavedStateWithArgs
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import java.time.LocalDateTime
+import java.time.ZoneOffset
+import java.util.Date
+
+class VariationDetailViewModelTest : BaseUnitTest() {
```
Comment (wzieba) [inline WooCommerce/src/test/kotlin/com/woocommerce/android/ui/products/variations/VariationDetailViewModelTest.kt:25]: This test is has a minimal setup to test code changed in scope of this PR

File: WooCommerce/src/test/kotlin/com/woocommerce/android/ui/products/variations/VariationDetailViewModelTest.kt [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        assertThat(variation?.saleStartDateGmt).isEqualTo(SALE_START_DATE)
+    }
+
+    @Test
+    fun `should update sale end date to null when requested`() {
+        sut.variationViewStateData.observeForever { _, _ -> }
+
+        sut.onVariationChanged(saleEndDate = null)
+
+        assertThat(variation!!.saleEndDateGmt).isNull()
```
Comment (wzieba) [inline WooCommerce/src/test/kotlin/com/woocommerce/android/ui/products/variations/VariationDetailViewModelTest.kt:93]: Those `!!` are intended - I want to make sure that we test `null` of `saleEndDateGmt`, not `null` that could be a result of null `variation`

**Review by 0nko — COMMENTED** [review]
The sale dates are looking good! But I'm experience some weird behavior with variation images:

1. Add a new image to a variation with no image
2. Wait for the upload to finish
3. Update
4. So far so good
5. Go back to the gallery and remove the image
6. Go back to the detail and notice there's a placeholder now, which shouldn't be there
7. Tap on the placeholder and see it shown in the gallery
8. Tap on it and notice the error snackbar

image
image

**Comment by wzieba** [comment]
Thanks for checking that, I confirm that I can reproduce but I'm also having the same experience on `develop` (it seems that this PR doesn't introduce regression).

Could you please confirm that it's happening also on `develop`?  If so, I'll create an issue for that and I think we can handle this bug in separate PR, WDYT?

Reproduction video from app build on `develop`

https://user-images.githubusercontent.com/5845095/112603023-b27da180-8e14-11eb-93e4-1f595ff346d6.mov

**Comment by 0nko** [comment]
I can reproduce it on `develop` as well. It's got to be something recent, though, because I don't see this in a build from a different branch from March 18 (ed904c34e2bdfc302311d46ee8268d1187e0a75d). 

Let's create a separate issue 👍.

**Review by 0nko — APPROVED** [review]
:shipit:

**Comment by wzieba** [comment]
Thanks! I've reported it: #3755

**Pull request merged**